proposals/2024-09-25_metadata-labels.md

```diff
@@ -0,0 +1,154 @@
+
+### Pitfalls of the current solution
+
+Roughly half of OpenTelemetry users surveyed preferred keeping metric names unmodified when translating to Prometheus. With our goal of being the default choice to store OpenTelemetry metrics, we should find a way to preserve metric names without introducing potential issues for users.
+
+## Goals
+
+Goals and use cases for the solution as proposed in [How](#how):
+
+* [Required] OpenTelemetry users can query for the original names of their metrics.
+* [Required] Users can filter by the metric type or unit in PromQL queries.
+* [Required] PromQL returns a warning when querying across metrics with different units or types, or when using "inappropriate" operations for a metric type.
+* [Nice to have] OpenTelemetry users can grep for the original names in the text exposition.
+
+### Audience
+
+This document is for Prometheus server maintainers, PromQL maintainers, and anyone interested in furthering compatibility between Prometheus and OpenTelemetry.
+
+## Non-Goals
+
+* Prometheus will not attempt to auto-convert between units when there is a conflict.
+* Prometheus will not attempt to auto-convert between types (e.g. native histogram vs float series).
+* Prometheus client libraries will not allow mixing metrics with the same name, but different type or unit in the exposition format. See potential future extensions.
+
+## User Experience
+
+When querying for a metric, users can filter for a type or unit by specifying a filter on the `__unit__` or `__type__` labels, which use the reserved `__` prefix to ensure they do not collide with user-provided labels.
+
+For example:
+
+* `my_metric{}` returns all series with any type or unit, including `__type__` and `__unit__` labels.
+* `my_metric{__unit__="seconds", __type__="counter"}` returns only series with the specified type and unit.
+
+When a query for a metric returns multiple metrics with a different `__type__` or `__unit__` label, but the same `__name__`, users see a warning in the UI.
+
+Users don't see the `__type__` or `__unit__` labels in the Prometheus UI next to other labels, but the unit displayed next to the value.
+
+Users see no difference to exposition formats.
+
+## How
+
+### PromQL Changes
+
+When a query for a metric returns multiple metrics with a different `__type__` or `__unit__` label, but the same `__name__`, an info annotation will be returned with the PromQL response, which is otherwise unmodified.
+
+### Prometheus UI Changes
+
+When displaying a metric's labels in the table or in the graph views, the UI will hide labels starting with `__` (double underscore) by default, similar to the current handling of `__name__`. A "Show System Labels" check-box will be added, which shows hidden labels when checked.
+
+### Prometheus Server Ingestion
+
+When receiving OTLP or PRW 2.0, or when scraping the text, OM, or proto formats, the type and unit of the metric are added as the `__type__` and `__unit__` labels, if not already present.
```

Comment:
I was thinking about what would be on the `/federate` endpoint if the Prometheus server scraped conflicting metrics from applications. Our choices are:

1. [current] Only `#UNIT` metadata, and no `__unit__` labels.
2. Only `__unit__` labels, and no `#UNIT` metadata.
3. [proposal] Mixed `__unit__` labels, and a `#UNIT` with one of the units the metric has.

* 1 will result in collisions, but maybe that is acceptable for the `/federate` endpoint.  It isn't a regression from what we have today.
* 2 works, but would be breaking for existing users that are consuming `#UNIT`.
* 3 is confusing because of the mixing, but I wanted to leave that open as an option.

I'm OK with disallowing mixing `__unit__` and `#UNIT`, since we could relax that in the future if needed.

Reply:
We discussed offline, and:

1. would epic to follow 1 and allow multiple the same named metric for different types and units in OM, let's try proposing that, it would help massively with exporters and federate
2. type and unit from metadata is primary (in case of mix)
3. Should we allow __type__ and __unit__ at all in exposition, SHOULD NOT maybe?

Reply:
I'm not too fond of adding __unit__ and __type__ as labels in the exposition format. I think they should stay as a single line at the top of the metric family like it is today.

The Unit/Type will be the same for all time series in a metric family. If they become labels in the exposition format, it will just be repeated information for all time series. It increases the payload, and we waste CPU time parsing it.

Reply:
We have "Users see no difference to exposition formats." stated above, but if you put the type and especially the unit in the `# UNIT` part then you need them in the metric name as well, defeating the purpose. You're other option is to put them into the series on each line.

Looking at /metrics is for humans and follows what you see is what you get in terms of showing what you can query - implying that if you can query by `__type__` or the API/UI returns the `__type__` then you should put them on each line.

And the implication of that is that there needs to be at least one more efficient exposition format that is not so verbose for machines to read.

So I'm not sure this would work without either putting the new labels on every line or modifying OpenMetrics.

Reply:
Of course the above argument doesn't stand if we follow @beorn7 's advice and not show `__type__`, `__unit__` on the query API (`/query` , `/query_range`) , although I would love to have it on the metadata queries endpoint like (`/series`) for Grafana to be able to know what's a float series and what's a native histogram.

Reply:
I think we should show `__type__` and `__unit__` in the query APIs. I'm just saying, in the zeroth step, we should remove those labels in PromQL operations (like the name is removed).

Reply:
I agree with showing `__type__` and `__unit__` in query APIs, just like we show `__name__`.

> We have "Users see no difference to exposition formats." stated above, but if you put the type and especially the unit in the # UNIT part then you need them in the metric name as well, defeating the purpose. You're other option is to put them into the series on each line.

That's a really good catch! This is something we'll probably need to tackle with OM 2.0. Is this what you had in mind with this issue @bwplotka https://github.com/prometheus/OpenMetrics/issues/280?

Reply:
By "Users see no difference to exposition formats", I meant that this proposal does not change how metrics are currently exposed in exposition formats.  I'll leave that for follow-up discussions in https://github.com/prometheus/OpenMetrics/issues/280 and elsewhere. I updated the language in https://github.com/prometheus/proposals/pull/39/commits/0a545c8488c99eaa83ae6a44df9636896c71753c.

I am open to ways of querying for type and unit that align better with the existing #TYPE and #UNIT comments, if we want to preserve the "what you see is what you get" experience of Prometheus.  We do support querying for `__name__`, so there is at least some precedence.